wallet/invoices.c

```diff
@@ -0,0 +1,402 @@
+	void *cbarg;
+};
+
+struct invoices {
+	/* The database connection to use. */
+	struct db *db;
+	/* The log to report to. */
+	struct log *log;
+	/* The invoice list. */
+	struct list_head invlist;
+	/* Waiters waiting for any new invoice to be paid. */
+	struct list_head waitany_waiters;
+};
+
+static void trigger_invoice_waiter(struct invoice_waiter *w,
+				   struct invoice *invoice)
+{
+	w->triggered = true;
+	w->cb(invoice, w->cbarg);
+}
+
+static bool wallet_stmt2invoice(sqlite3_stmt *stmt, struct invoice *inv)
+{
+	inv->id = sqlite3_column_int64(stmt, 0);
+	inv->state = sqlite3_column_int(stmt, 1);
+
+	assert(sqlite3_column_bytes(stmt, 2) == sizeof(struct preimage));
+	memcpy(&inv->r, sqlite3_column_blob(stmt, 2), sqlite3_column_bytes(stmt, 2));
+
+	assert(sqlite3_column_bytes(stmt, 3) == sizeof(struct sha256));
+	memcpy(&inv->rhash, sqlite3_column_blob(stmt, 3), sqlite3_column_bytes(stmt, 3));
+
+	inv->label = tal_strndup(inv, sqlite3_column_blob(stmt, 4), sqlite3_column_bytes(stmt, 4));
+
+	if (sqlite3_column_type(stmt, 5) != SQLITE_NULL) {
+		inv->msatoshi = tal(inv, u64);
+		*inv->msatoshi = sqlite3_column_int64(stmt, 5);
+	} else {
+		inv->msatoshi = NULL;
+	}
+
+	inv->expiry_time = sqlite3_column_int64(stmt, 6);
+	/* Correctly 0 if pay_index is NULL. */
+	inv->pay_index = sqlite3_column_int64(stmt, 7);
+
+	if (inv->state == PAID)
+		inv->msatoshi_received = sqlite3_column_int64(stmt, 8);
+
+	list_head_init(&inv->waitone_waiters);
+	return true;
+}
+
+struct invoices *invoices_new(const tal_t *ctx,
+			      struct db *db,
+			      struct log *log)
+{
+	struct invoices *invs = tal(ctx, struct invoices);
+
+	invs->db = db;
+	invs->log = log;
+
+	list_head_init(&invs->invlist);
+	list_head_init(&invs->waitany_waiters);
+
+	return invs;
+}
+
+
+bool invoices_load(struct invoices *invoices)
+{
+	int count = 0;
+	struct invoice *i;
+	sqlite3_stmt *stmt;
+
+	/* Load invoices from db. */
+	stmt = db_query(__func__, invoices->db,
+			"SELECT id, state, payment_key, payment_hash"
+			"     , label, msatoshi, expiry_time, pay_index"
+			"     , msatoshi_received"
+			"  FROM invoices;");
+	if (!stmt) {
+		log_broken(invoices->log, "Could not load invoices");
+		return false;
+	}
+
+	while (sqlite3_step(stmt) == SQLITE_ROW) {
+		i = tal(invoices, struct invoice);
+		if (!wallet_stmt2invoice(stmt, i)) {
+			log_broken(invoices->log, "Error deserializing invoice");
+			sqlite3_finalize(stmt);
+			return false;
+		}
+		list_add_tail(&invoices->invlist, &i->list);
+		count++;
+	}
+	log_debug(invoices->log, "Loaded %d invoices from DB", count);
+
+	sqlite3_finalize(stmt);
+	return true;
+}
+
+const struct invoice *invoices_create(struct invoices *invoices,
+				      u64 *msatoshi TAKES,
+				      const char *label TAKES,
+				      u64 expiry)
+{
+	sqlite3_stmt *stmt;
+	struct invoice *invoice;
+	struct preimage r;
+	struct sha256 rhash;
+	u64 expiry_time;
+
+	if (invoices_find_by_label(invoices, label)) {
```

Comment:
This is a duplicate to the test in `json_invoice`, but probably a good idea to be keep both, since the other one has a nice error message attached to it. The downside is that we now may return `NULL` to the caller.

Reply:
Agree.  Mostly the principle is that the invoices system could be ripped out and given a different interface other than the RPC (currently the invoices system has only the RPC as interface, but perhaps c-lightning could become e.g. a library that can be called from e.g. a high-level language like Python --- basically one reason for separating the invoices system is the potential for more flexibility later. Of course, one can also argue that it is overdesign instead.)